packages/mco/components/discovered-application-wizard/wizard-steps/namespace-step/namespace-step.tsx

```diff
@@ -44,17 +58,121 @@ const clusterOptions = (clusters: string[]): JSX.Element[] =>
     />
   ));
 
+const getInputValidationSchema = (t: TFunction, existingNames: string[]) => {
+  const fieldRequirements = [
+    fieldRequirementsTranslations.maxChars(t, 63),
+    fieldRequirementsTranslations.startAndEndName(t),
+    fieldRequirementsTranslations.alphaNumericPeriodAdnHyphen(t),
+    fieldRequirementsTranslations.cannotBeUsedBefore(t),
+  ];
+
+  const schema = Yup.object({
+    'name-input': Yup.string()
+      .required()
+      .max(63, fieldRequirements[0])
+      .matches(
+        validationRegEx.startAndEndsWithAlphanumerics,
+        fieldRequirements[1]
+      )
+      .matches(
+        validationRegEx.alphaNumericsPeriodsHyphensNonConsecutive,
+        fieldRequirements[2]
+      )
+      .test(
+        'unique-name',
+        fieldRequirements[3],
+        (value: string) => !existingNames.includes(value)
+      ),
+  });
+
+  return { schema, fieldRequirements };
+};
+
+export const NameInput: React.FC<NameInputProps> = ({
+  name,
+  drPlacements,
+  dispatch,
+}) => {
+  const { t } = useCustomTranslation();
+
+  const { schema, fieldRequirements } = React.useMemo(() => {
+    const existingNames =
+      drPlacements
+        ?.filter(
+          (drpc) => getNamespace(drpc) === RAMEN_PROTECTED_APPS_NAMESPACE
+        )
+        .map(getName) || [];
+    return getInputValidationSchema(t, existingNames);
+  }, [drPlacements, t]);
+
+  const resolver = useYupValidationResolver(schema);
+  const {
+    control,
+    watch,
+    formState: { isValid },
+  } = useForm({
+    ...formSettings,
+    resolver,
+  });
+
+  const newName: string = watch('name-input');
+
+  React.useEffect(
+    () =>
+      dispatch({
+        type: EnrollDiscoveredApplicationStateType.SET_NAME,
+        payload: isValid ? newName : '',
+      }),
+    [newName, isValid, dispatch]
+  );
```

Comment:
```suggestion
  React.useEffect(
    () =>
      dispatch({
        type: EnrollDiscoveredApplicationStateType.SET_NAME,
        payload: newName,
      }),
    [newName, isValid, dispatch]
  );
```
couple of things here:
1. I don't think it is good UX to automatically reset the state when input is invalid. We can disable the "Next" button and show warning/error messages... but should not reset input to `'' (empty string)` automatically.
(let me know if I misunderstood and that was not the intention here).

2. We can eliminate use of `useEffect` entirely IMO (it is good practice to eliminate useEffect where ever possible). `textInputProps` takes `onChange` prop and we can directly pass `dispatch` there. I know other places we are using `useEffect` for this, but to me it seems like a miss and overkill.
(let me know if it does not works, or there is any issue with this approach).

Reply:
TextInputWithFieldRequirements.tsx is maintaining its own state, Even though I am setting an empty string, it won't disappear from the UI. Only validation fails.

Let me do one thing, I will store the isValid also in state so I can use the it for the validation

Reply:
`I will store the isValid also in state so I can use the it for the validation` >> why, do we need it ?? I don't think so.

Reply:
useEffect is to update my wizard page state

Reply:
u can directly dispatch from `onChange` of `textInputProps`, instead of doing it from `useEffect` ? that will update ur wizard page...
```
onChange = {() =>  dispatch({
        type: EnrollDiscoveredApplicationStateType.SET_NAME,
        payload: newName,
      })}
```

Reply:
Text input is displayed from TextInputWithFieldRequirements state,  if I clean my state, it will still display the text in the input box, I am taking advantage of this behavior,

When isValid is false, my state is cleaned, So validation fails, if the user clicks the next button they will see an error message, But still, the text is displayed in the input box from the TextInputWithFieldRequirements state.

Reply:
regarding `onChange` part: https://react.dev/learn/you-might-not-need-an-effect#how-to-remove-unnecessary-effects (`You don’t need Effects to handle user events`)...

Reply:
I checked onChange behaviour, it return the input text but isValid is an async updated value, it is not changing at onChange, It looks  useEffect is used on purpose.

Reply:
ack !